MA-199 Course Authorization framework in mobile API

nasthagiri opened the pull request:
Please review.  @aleffert @BenjiLee

aleffert commented on lms/djangoapps/courseware/access.py:
```diff
@@ -136,6 +140,36 @@ def can_load_forum():
+            # check enrollment
+            (
+                CourseEnrollment.is_enrolled(user, course.id) or
+                _has_staff_access_to_descriptor(user, course, course.id)
```
I wonder if there's a way we can make the names more obvious. Since it's not just enrollment, it's also beta users though IIRC I had the same problem and didn't come up with anything good.

nasthagiri replied:
The `can_load_mobile` method checks enrollment in addition to the other checks.
The `can_load_mobile_not_enrolled` method does not check enrollment, but checks release date, beta user, and eventually cohorted content (once supported).

How about 'can_load_mobile' and 'can_load_mobile_no_enrollment_check'?

aleffert replied:
That sounds good to me.

nasthagiri replied:
Renamed load_mobile_not_enrolled -> load_mobile_no_enrollment_check in latest commit.

aleffert commented on lms/djangoapps/mobile_api/course_info/tests.py:
```diff
@@ -9,35 +9,40 @@
-    Tests for /api/mobile/v0.5/course_info/...
+    Base test class for tests for /api/mobile/v0.5/course_info/...
     """
-    def setUp(self):
```
What's the idea behind an empty base class?

nasthagiri replied:
No.  It's so all subclasses can share a common hierarchy and implicitly inherit from `MobileCourseAPITestCase`.
Also, I was thinking there may be common `setUp` required, but eventually didn't need it.

nasthagiri replied:
I've removed it in my latest commit.

aleffert commented on lms/djangoapps/mobile_api/utils.py:
```diff
@@ -1,31 +1,92 @@
     """
+    def _decorator(func):
+        """Outer method decorator."""
+        @functools.wraps(func)
```
Maybe this isn't a big deal, but I believe (and my python is not amazing so I'm not totally sure) that this will make a function that claims it takes an argument called ``course``, but actually takes an argument called ``course_id``?

nasthagiri replied:
The `functools.wraps` decorator doesn't override the declaration of the argument list.
Per its doc (see `update_wrapper` at https://docs.python.org/3.1/library/functools.html)
and its code (https://github.com/enthought/Python-2.7.3/blob/master/Lib/functools.py#L16), it only overrides the following attributes of the function: `'__module__' (name of the module the function is defined in), '__name__' (function name), '__doc__' (docstring), and '__dict__' (attributes set on the function)`.

aleffert commented on lms/djangoapps/mobile_api/users/tests.py:
```diff
@@ -163,154 +94,134 @@ def _setup_course_skeleton(self):
-        json_data = json.loads(result.content)
-        self.assertEqual(result.status_code, 400)
-        self.assertEqual(json_data, errors.ERROR_INVALID_MODULE_ID)
+        response = self.get_response(data={"last_visited_module_id": unicode(other_unit.location)})
```
I found it kind of confusing that this thing was called ``get...`` but the HTTP method was overridden so it was actually a PATCH. Maybe there's a better name so it's more obvious whether it's it's get like a getter or get like a GET request.

nasthagiri replied:
How about `api_response`?

aleffert commented on lms/djangoapps/mobile_api/tests.py:
```diff
@@ -21,35 +26,233 @@
+        return self.client.get(url)
+
+
+class MobileCourseAPITestCase(MobileAPITestCase):
```
I found the proliferation of mixins with a deep class hierarchy kind of confusing here and it makes it harder to just compose things.

My naive thought would be that something that had a course and a user argument would mix in the course test case and the user test case, but how it is here, they both bring in a whole bunch of (the same) functionality and it requires having a third custom combination. I imagine we'll have a few more types that are super common and having to have a mixin for each combination seems excessive.

As a proposal, what do you think of the idea of making ``reverse_args`` a method that can be called with decorators that add to its output.

So you wouldn't need e.g. MobileUserCourseAPITestCase. There would just be just a ``user_argument`` class decorator and a ``course_argument`` class decorator. Each of these would append the relevant thing to the result of a ``get_reverse_args`` method that as a stub returns an empty dict.

For extra convenience if we wanted it, we could have MobileUserCourseAPITestCase then be the class that inherits from ``MobileAPITestCase`` with both of those decorators.

nasthagiri replied:
I was thinking afterwards I should have added further documentation/description about the class and mixin hierarchy.

First, the rationale for test class versus test mixin:
1. To prevent the same test method from being run multiple times for a given test class, all inherited test methods are defined in mixins - which generally do not have a hierarchy - and are explicitly included by concrete test classes.   (Alternatively, if a test was implemented in a common base class inherited multiple times by a test class, it would be run multiple times).

2. To keep the mixin usage simpler (and avoid multiple inheritance issues), each concrete test class simply includes the appropriate test mixin that applies to its endpoint.

So here's the resulting class and mixin hierarchy:
MobileAPITestCase
  <-- MobileCourseAPITestCase
  <-- MobileUserAPITestCase
  <-- MobileUserCourseAPITestCase

MobileAuthTestMixin *(is_user=False)*
  <-- MobileAuthUserTestMixin
MobileCourseAccessTestMixin *(verify_enrolled=False)*
  <-- MobileEnrolledCourseAccessTestMixin

nasthagiri replied:
Now, given the above, I like your suggestion of removing the class hierarchy under `MobileAPITestCase` (although note, it's not a deep hierarchy but rather a shallow hierarchy of depth 1).

I'll have to think further whether/how it could work with decorators.

nasthagiri replied:
After having thought about this further, even with decorators for the reverse_args function, each concrete test class would still need to declare the correct decorator(s) - which could be DRY-ed with a base class -- which ends up where we are right now.

Also, having reread your 2nd paragraph above, I believe you may have misunderstood the class/mixin structure.  Perhaps my description of the class/mixin above explains how it does *not* "bring in a whole bunch of (the same) functionality".  There is no requirement "to have a mixin for each combination".  In fact, it is expected that each concrete class include the appropriate mixins directly.

But if there's a good way of eliminating the proliferation of classes for the various implementations of `reverse_args` or as currently done, of `get_url`, that would be good.

nasthagiri replied:
Ok.  I think I've addressed your concern with the latest commit by introducing a `REVERSE_INFO` dict that all concrete test classes must implement.  I believe it makes the test class hierarchy design simpler by removing all the intermediary subclasses.  We now only have 1 base test class and individual Test Mixins.

aleffert commented:
That's all from me.

BenjiLee commented on lms/djangoapps/mobile_api/tests.py:
```diff
@@ -3,15 +3,20 @@
+from student.models import CourseEnrollment
 
 from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
 from xmodule.modulestore.tests.factories import CourseFactory
```
Duplicate imports.

nasthagiri replied:
Oops! Thanks.  I'm surprised pylint doesn't raise this issue.

BenjiLee commented on lms/djangoapps/mobile_api/tests.py:
```diff
@@ -21,35 +26,233 @@
+        self.init_course_access()
+
+        response = self.get_response(expected_response_code=None)
+        self.verify_success(response)
```
Small nitpick; I think we are checking for 200 response twice. Doesn't the get_response() check for 200 response by default?

nasthagiri replied:
In this case, no, since `expected_response_code=None`.  This allows subclasses to override the implementation of this.

BenjiLee commented on lms/djangoapps/mobile_api/tests.py:
```diff
@@ -21,35 +26,233 @@
+        self.init_course_access(course_id=non_existent_course_id)
+
+        response = self.get_response(expected_response_code=None, course_id=non_existent_course_id)
+        self.verify_failure(response)
```
To add to the last comment about get_response(), wouldn't setting expected_response_code=404 do the same thing as self.verify_failure()?

nasthagiri replied:
Calling `self.verify_failure` allows subclasses to handle the failure case.  For example, `TestUserEnrollmentApi` overrides `verify_failure` since it always returns a `200`, but with `0` courses.

BenjiLee commented on lms/djangoapps/mobile_api/users/tests.py:
```diff
@@ -3,149 +3,80 @@
 
 import datetime
-import ddt
 import json
```
Unused import

nasthagiri replied:
Hmm.. once again.. I'm surprised this wasn't caught by pylint.  Thanks!

nasthagiri commented on lms/djangoapps/mobile_api/utils.py:
```diff
@@ -1,31 +1,92 @@
 
 
-def mobile_available_when_enrolled(course, user):
+def mobile_course_access(depth=0, verify_enrolled=True):
```
It is being used by `VideoSummaryList` here: https://github.com/edx/edx-platform/pull/6398/files#diff-4bef877c8b70416409d4e1c5021cf45cR82

nasthagiri commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -274,38 +234,22 @@ class UserCourseEnrollmentsList(generics.ListAPIView):
-        return mobile_course_enrollments(qset, self.request.user)
+        enrollments = self.queryset.filter(user__username=self.kwargs['username'], is_active=True).order_by('created')
+        return [
+            enrollment for enrollment in enrollments
```
Yeah, generally list comprehensions are the preferred conventional way in python, with `filter` possibly becoming deprecated.  But I could see that it could be a matter of style.
I continue to use list comprehensions for filtering since the expression makes it explicit that we're creating a list.

nasthagiri commented:
@BenjiLee @aleffert  I believe I have addressed all your comments.  Please see the latest commit https://github.com/edx/edx-platform/commit/a2217274d63911d377a5762c1043891147b3f100.

BenjiLee commented on lms/djangoapps/mobile_api/course_info/tests.py:
```diff
@@ -9,35 +9,33 @@
+from ..testutils import (
+    MobileAPITestCase, MobileCourseAccessTestMixin, MobileEnrolledCourseAccessTestMixin, MobileAuthTestMixin
+)
 
```
More unused imports. Seems like pylint is not picking these up.

`import json
from django.core.urlresolvers import reverse
from rest_framework.test import APITestCase

from courseware.tests.factories import UserFactory

from xmodule.modulestore.tests.factories import CourseFactory`

BenjiLee commented on lms/djangoapps/mobile_api/course_info/views.py:
```diff
@@ -6,13 +6,15 @@
 from static_replace import make_static_urls_absolute, replace_static_urls
 
+from ..utils import MobileView, mobile_course_access
 
```
More unused imports: 
`from rest_framework import permissions`
`from rest_framework.authentication import OAuth2Authentication, SessionAuthentication`

`from opaque_keys.edx.keys import CourseKey`
`from xmodule.modulestore.django import modulestore`

aleffert commented:
:+1:

BenjiLee commented on lms/djangoapps/mobile_api/testutils.py:
```diff
@@ -0,0 +1,204 @@
+        self.client.login(username=self.username, password=self.password)
+
+    def logout(self):
+        """Login test user."""
```
Logout*

nasthagiri replied:
fixed.

BenjiLee commented:
:+1:
